chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+     )
+
+     ;; utility to fetch coin amount from coin
+     (defun-inline input_amount_for_coin (coin)
+       (f (r (r coin)))
+     )
+
+     ;; calculate the hash of an announcement
+     (defun-inline calculate_annoucement_id (this_coin_id this_subtotal next_coin_id cat_mod_struct)
+       ; NOTE: the next line containts a bug, as sha256tree1 ignores `this_subtotal`
+       (sha256 next_coin_id 0xcb this_coin_id this_subtotal)
+     )
+
+     ;; create the `ASSERT_COIN_ANNOUNCEMENT` condition that ensures the next coin's announcement is correct
+     (defun-inline create_assert_next_announcement_condition (this_coin_id this_subtotal next_coin_id cat_mod_struct)
+       (list ASSERT_COIN_ANNOUNCEMENT
+             (calculate_annoucement_id this_coin_id
+                                            this_subtotal
+                                            next_coin_id
+                                            cat_mod_struct
+             )
+       )
+     )
+
+     ;; here we commit to I_{k-1} and S_k
+     (defun-inline create_announcement_condition (prev_coin_id prev_subtotal)
+       (list CREATE_COIN_ANNOUNCEMENT
+             (concat 0xcb (concat prev_coin_id prev_subtotal))
+       )
+      )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+     ;; this function takes a condition and returns an integer indicating
+     ;; the value of all output coins created with CREATE_COIN. If it's not
+     ;; a CREATE_COIN condition, it returns 0.
+
+     (defun-inline output_value_for_condition (condition)
+       (if (= (f condition) CREATE_COIN)
+         (f (r (r condition)))
+         0
+       )
+     )
+
+     ;; this function takes a list of conditions and returns an integer indicating
+     ;; the value of all output coins created with CREATE_COIN
+     (defun output_totals (conditions)
+       (if conditions
+         (+ (output_value_for_condition (f conditions)) (output_totals (r conditions)))
+         0
+       )
+     )
+
+     ;; add three conditions to the list of morphed conditions:
+     ;; ASSERT_MY_COIN_ID for `this_coin_info`
+     ;; CREATE_COIN_ANNOUNCEMENT for my announcement
+     ;; ASSERT_COIN_ANNOUNCEMENT for the next coin's announcement
+     (defun-inline generate_final_output_conditions
+       (
+         prev_subtotal
+         this_subtotal
+         morphed_conditions
+         prev_coin_id
+         this_coin_id
+         next_coin_id
+         cat_mod_struct
+       )
+         (c (create_announcement_condition prev_coin_id prev_subtotal)
+           (c (create_assert_next_announcement_condition this_coin_id this_subtotal next_coin_id cat_mod_struct)
+             morphed_conditions)
+         )
+      )
+
+    (defun-inline coin_info_for_coin_bundle (coin_bundle)
+      (f coin_bundle)
+    )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;; lineage checking
+
+     ;; return true iff parent of `this_coin_info` is provably a cat
+     (defun is_parent_cat (
```

Comment:
I don't quite understand... we either fail or return `1`... so the return value is never used? This seems like a weird construct.

Reply:
Looks like this is only called in one place... consider moving the `if` portion to the caller, so this actually returns a boolean. (Functions named `is_*` should generally return a boolean. Raising is a very surprising behaviour to me.)

Reply:
Changed to return ()